cli/internal/runsummary/spaces.go

```diff
@@ -1,15 +1,259 @@
+	// so we have to construct these URLs lazily with a `makeURL` affordance.
+	//
+	// We are assuming that if makeURL is defined, this is NOT the first request.
+	// This is not a great assumption, and will fail if our endpoint URLs change later.
+	//
+	// Secondly, if makeURL _is_ defined, we call it, and if there are any errors, we exit early.
+	// We are doing this check before any of the other basic checks (e.g. the existence of a spaceID)
+	// becaus in the case the repo is not linked to a space, we don't want to print those errors
+	// for every request that fails. On the other hand, if that POST /run request fails, and N
+	// requests fail after that as a consequence, it is ok to print all of those errors.
+	if req.makeURL != nil {
+		if err := req.makeURL(req, c.run); err != nil {
+			c.errors = append(c.errors, err)
+			return
+		}
+	}
+
+	if c.spaceID == "" {
+		c.errors = append(c.errors, req.error("No spaceID found"))
+		return
+	}
+
+	if !c.api.IsLinked() {
+		c.errors = append(c.errors, req.error("Repo is not linked to a Space. Run `turbo link --target=spaces` first"))
+		return
+	}
+
+	// We only care about POST and PATCH right now
+	if req.method != "POST" && req.method != "PATCH" {
+		c.errors = append(c.errors, req.error(fmt.Sprintf("Unsupported method %s", req.method)))
+		return
+	}
+
+	payload, err := json.Marshal(req.body)
+	if err != nil {
+		c.errors = append(c.errors, req.error(fmt.Sprintf("Failed to create payload: %s", err)))
+		return
+	}
+
+	// Make the request
+	var resp []byte
+	var reqErr error
+	if req.method == "POST" {
+		resp, reqErr = c.api.JSONPost(req.url, payload)
+	} else if req.method == "PATCH" {
+		resp, reqErr = c.api.JSONPatch(req.url, payload)
+	} else {
+		c.errors = append(c.errors, req.error("Unsupported request method"))
+	}
+
+	if reqErr != nil {
+		c.errors = append(c.errors, req.error(fmt.Sprintf("%s", reqErr)))
+		return
+	}
+
+	// Call the onDone handler if there is one
+	if req.onDone != nil {
+		req.onDone(req, resp)
+	}
+}
+
+func (c *spacesClient) createRun(rsm *Meta) {
+	if !c.enabled {
+		return
+	}
+
+	c.queueRequest(&spaceRequest{
+		method: "POST",
+		url:    fmt.Sprintf(runsEndpoint, c.spaceID),
+		body:   newSpacesRunCreatePayload(rsm),
+
+		// handler for when the request finishes. We set the response into a struct on the client
+		// because we need the run ID and URL from the server later.
+		onDone: func(req *spaceRequest, response []byte) {
+			if err := json.Unmarshal(response, c.run); err != nil {
+				c.errors = append(c.errors, req.error(fmt.Sprintf("Error unmarshaling response: %s", err)))
+			}
+
+			// close the run.created channel, because all other requests are blocked on it
+			close(c.runCreated)
+		},
+	})
+}
+
+func (c *spacesClient) postTask(task *TaskSummary) {
+	if !c.enabled {
+		return
+	}
+
+	c.queueRequest(&spaceRequest{
+		method: "POST",
+		makeURL: func(self *spaceRequest, run *spaceRun) error {
+			if run.ID == "" {
```

Comment:
Are you ever sending this request when there's no `run.ID`? I guess that would happen if the first request failed? 

There might be a better way to handle that scenario (checking `c.run.ID` after `c.runCreated` closes). Then, by the time you are calling `dequeueRequest`, you already know whether or not to expect an id to be present, and can short-circuit there.

Reply:
I can do that, but I'd need that additional bookkeeping, and it would be "far" enough away from this makeURL function that I'd want this check here anyway.